### DISCUSSION TITLE
Serialization of type object is not supported while using property as complex type in Business model.

### DISCUSSION THREAD

**Discussion by Hvishal88** [MarimerLLC/csla#3545] [Questions]
Getting an error "Serialization of type object is not supported" while using property type as non-primitive datatype (e.g. Class).

Any solution to get this serialized or ignored during DAL operations.

```c#
    // POST: Person/Edit/5
    [HttpPost]
    [ValidateAntiForgeryToken]
    public async Task<ActionResult> Edit(int id, PersonEdit person)
    {
      try
      {
        **person.ComplexClassProp = new ComplexClass() { FirstName = "Vishal" };**
        LoadProperty(person, PersonEdit.IdProperty, id);
       
        if (await SaveObjectAsync<PersonEdit>(person, true))
          return RedirectToAction(nameof(Index));
        else
          return View(person);
      }
      catch
      {
        return View(person);
      }
    }
```

-------------------------------------------------------------

```c#
 public class ComplexClass
  {
    public string FirstName { get; set; }
    public string LastName { get; set; }
  }
  [Serializable]
  public class PersonEdit : BusinessBase<PersonEdit>
  {

    public static readonly PropertyInfo<ComplexClass> ComplexClassPropProperty = RegisterProperty<ComplexClass>(nameof(ComplexClassProp));
    


    public ComplexClass ComplexClassProp
    {
      get { return GetProperty(ComplexClassPropProperty); }
      set { SetProperty(ComplexClassPropProperty, value); }
    }
```
.
.
.
.
.
Find the attached sample code for the issue reference. 
[MvcExample.zip](https://github.com/MarimerLLC/csla/files/13297132/MvcExample.zip)

**Comment by hurcane** [comment]
I think you need to add the Serialiazble attribute to ComplexClass.

**Comment by rockfordlhotka** [comment]
The `MobileFormatter` is used to serialize business object graphs in CSLA. All types referenced by a domain class must be serializable type `MobileFormatter`.

`MobileFormatter` knows how to serialize "primitive" types (actual primitive, plus `string`, date types, and a few others) and types that implement `IMobileObject`. It can not serialize arbitrary types - partially because that'd be a massive security hole - and also because it isn't possible (even with reflection) to serialize and deserialize all types that can be implemented by C#.

In your case, the easiest thing is to implement `SimpleClass` as a subclass of `Csla.Core.MobileObject`. This is the simplest implementation of `IMobileObject`, though it does require that you do some extra work.

Typed from memory, so hopefully this is correct:

```c#
  public class SimpleClass : Csla.Core.MobileObject
  {
    public string FirstName { get; set; }
    public string LastName { get; set; }

    protected override void OnGetState(SerializationInfo info, MobileFormatter formatter)
    {
      info.AddValue(nameof(FIrstName), FirstName);
      info.AddValue(nameof(LastName), LastName);
    }

    protected override void OnSetState(SerializationInfo info, MobileFormatter formatter)
    {
      FirstName = info.GetValue<string>(nameof(FirstName));
      LastName = info.GetValue<string>(nameof(LastName));
    }
  }
```

Your other alternative is to subclass `BusinessBase`, though that carries more metastate and metadata in the serialization stream to support the rules engine and other advanced features of `BusinessBase`.
